### PR TITLE
[perf]: Eliminate CPU-GPU synchronization bottlenecks in training pipeline

### PR DESCRIPTION
**Pull request by rich7420**
## Purpose

Eliminate implicit `cudaDeviceSynchronize` stalls in the training loop that serialise the CPU command dispatcher
against the GPU execution engine, destroying compute/communication overlap during gradient accumulation.
Discovered via `nsys-ai profile_health_manifest` analysis of an Nsight Systems trace captured on a 2×L40S Modal run.

Fixes #

## Changes

**Problem identified by [nsys-ai](https://github.com/GindaChen/nsys-ai):**
The `profile_health_manifest` diagnostic flagged two root causes:
1. **Excessive Synchronization** — `.item()` calls on loss tensors inside the gradient accumulation loop forced a
   `cudaDeviceSynchronize` at every micro-batch, stalling the CPU until the GPU drained its entire command queue.
2. **Pageable Memory Transfers** — `tensor.to(device)` without `non_blocking=True` blocked the CPU thread until
   the PCIe DMA transfer completed, preventing concurrent kernel dispatch.
The original profile showed only **9.7% Compute/NCCL overlap** and **45.2% GPU idle time**.

### `fastvideo/training/training_pipeline.py`
- **Async H2D transfers:** Added `non_blocking=True` to all `.to(device)` calls for `latents`,
  `encoder_hidden_states`, and `encoder_attention_mask` in `_get_next_batch()`.
- **GPU-side loss accumulation:** Replaced `training_batch.total_loss += avg_loss.item()` with
  `training_batch.total_loss += avg_loss`, keeping the reduced loss tensor on GPU throughout
  gradient accumulation. The scalar is materialised once at the end via `float(training_batch.total_loss)`.
### `fastvideo/train/trainer.py`
- **Deferred scalar extraction:** Modified `_coerce_log_scalar()` to return detached GPU tensors
  (instead of calling `.item()`) during accumulation. Added a `force_cpu` flag for the final
  logging pass.
- **Single sync point:** Loss and metric dictionaries accumulate as GPU tensors. A single
  `float(v)` call per key occurs only after `optimizer.step()`, when the GPU is busy with the
  next iteration's forward pass.

## Test Plan

the test [files](https://github.com/rich7420/fastvideo/blob/test-profile/fastvideo/tests/modal/training_nsys_profile.py)

```bash
# 1. Functional validation (2×L40S on Modal)
modal run fastvideo/tests/modal/training_nsys_profile.py::run_profiler_sanity_check
# 2. Profile analysis
nsys-ai skill run profile_health_manifest mfu_training.sqlite
```

## Test Results

Modal training output (5 steps, 2×L40S, Wan2.1-T2V-1.3B)

<details>
<summary>Test output</summary>

```
# Paste output here
Steps: 100%|██████████| 5/5 [00:31<00:00, loss=0.0855, step_time=5.81s, grad_norm=0.854]
wandb: Run summary:
wandb: avg_step_time 6.36619
wandb:     step_time 5.81402
wandb:     grad_norm 0.85353
[profiler-sanity] Done. Exit code: 0
```
```
══ Profile Health Manifest ══
  GPU:          L40S
  Profile span: 50332.7ms

  Top Kernels:
    flash_bwd_dq_dk_dv_loop_seqk_parallel      6015.9ms  ×300
    ncclDevKernel_SendRecv                      4305.8ms  ×904
    ncclDevKernel_AllReduce_Sum_f32_RING_LL     3907.8ms  ×162
    flash_fwd_kernel                            3573.8ms  ×600
    elementwise_kernel                          1407.8ms  ×9090

  Compute/NCCL Overlap:
    Compute only: 21696.4ms
    NCCL only:    5861.2ms
    Overlap:      29.5%

  Root Causes (24 findings):
    🔴 NCCL Serialization
    🟡 GPU Bubbles (Pipeline Stalls)
    🟡 Excessive H2D Transfers
    🟡 Small Kernel Overhead
    🟡 Pipeline Imbalance

  ⚡ Suspected bottleneck: NCCL serialization (overlap < 30%)

```

</details>

## Checklist

- [x] I ran `pre-commit run --all-files` and fixed all issues
- [x] I added or updated tests for my changes
- [x] I updated documentation if needed
- [x] I considered GPU memory impact of my changes

**For model/pipeline changes, also check:**
- [x] I verified SSIM regression tests pass
- [ ] I updated the support matrix if adding a new model

### CODE REVIEW

**Comment by rich7420** [comment]
@SolitaryThinker thanks for the review!
